go.mod

```diff
@@ -142,9 +142,12 @@ require (
 	sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
-	tags.cncf.io/container-device-interface/specs-go v0.6.0 // indirect
+	tags.cncf.io/container-device-interface/specs-go v0.7.0 // indirect
 )
 
 // Workaround for indirect dependency no longer being available.
 // https://github.com/containerd/containerd/issues/9969
 exclude github.com/mitchellh/osext v0.0.0-20151018003038-5e2d6d41470f
+
+// We replace the v1.2.0 runtime-spec required by container-device-interface with v1.1.0.
+replace github.com/opencontainers/runtime-spec v1.2.0 => github.com/opencontainers/runtime-spec v1.1.0
```

Comment:
Curious; was there anything in runtime-spec 1.2.0 that's required by CDI, or does the repo have dependabot enabled, therefore automatically updating dependencies to "latest"?


The problem with a replace rule is that they only apply locally, and not to any consumer of the containerd module (i.e., anyone updating the containerd module will still be forced to update the runtime-spec).


If CDI itself does not require any of the features added in v1.1.0 or v1.2.0 of the spec, the correct approach may be to lower the required version (similar to https://github.com/containerd/platforms/pull/5, which was opened for this reason)

Go modules uses "Minimum Version Selection" (MVS); https://go.dev/ref/mod#minimal-version-selection. For _libraries_, this means that the `go.mod` should always contain the _minimum_ version needed to satisfy the library's requirements; which could be an older version of the runtime-spec (e.g. 1.0.0) of that version has all the required features. Specifying a low version means that _consumers_ of the module can decide what version to use, as long as it's SemVer compatible (go modules dictates that SemVer must be followed, so if the module requires (say) v1.0.0, then consumers are allowed to pick any v1.x.x version that they need).

In some cases, it may be needed to update minimum required version if a dependency contains a vulnerability in parts of the code that's used; but strictly only if the _code that's used_ is affected (this may be a bit of a grey area).

Dependabot can be a bit of an issue in these situations, as it directly conflicts with MVS principles when used on a _library_ module.

Reply:
There shouldn't be anything that we require from the `v1.2.0`, so we could definitely decrease the dependency there. We do have dependabot enabled in the repo and as you point out, we could look at disabling it instead.

Note that we do use `github.com/opencontainers/runtime-tools` to make OCI spec modifications and this has its own spec dependency: `https://github.com/opencontainers/runtime-tools/blob/master/go.mod#L11

This has also already caused issues in https://github.com/cri-o/cri-o/pull/7971 where the testing infrastructure (which uses `crun`) has not yet been updated to a crun version that supports a lower spec.

@klihub for your information.

Reply:
I have created https://github.com/cncf-tags/container-device-interface/pull/202 to decrease the dependency version.